### PR TITLE
Regex has non-deterministic behaviour

### PR DESCRIPTION
**Pull request by muxanick**
There were two issues:
**1.** RE2 returns null pointer (as start or end of match) and we implicitly
cast it to pointer and subtract other pointer, depends on a value of that
pointer a result could be positive or negative int value (depends on systems
and used memory). To fix that we can check for null pointer and assign
-1 to that pointer to create deterministic behaviour.
**2.** Sub-string throw exception if range borders are out of range, we want
to limit it.

### CODE REVIEW

**Comment by muxanick** [comment]
Fix for #739

File: javalib/src/main/scala/java/util/regex/Matcher.scala [hunk 1]
```diff
@@ -85,8 +85,8 @@ final class Matcher private[regex] (var _pattern: Pattern,
         var i = 0
         while (i < nMatches) {
           val m     = matches + i
-          val start = (m.data - in).toInt
-          val end   = start + m.length
+          val start = if (m.data == null) -1 else (m.data - in).toInt
+          val end   = if (m.data == null) -1 else start + m.length
```
Comment (densh) [inline javalib/src/main/scala/java/util/regex/Matcher.scala]: This seems to be like a workaround that doesn't address the root cause of the issue: we add `+ 1` to the number of groups in https://github.com/scala-native/scala-native/pull/740/files#diff-ef6e5cb9d8056c08c6441a0d16a64425R39, and then when we try to read the data of that last group it's simply not there (i.e. `m.data == null`.) Can we fix the root cause instead?

Comment (densh) [inline javalib/src/main/scala/java/util/regex/Matcher.scala]: @MasseGuillaume Do you remember why `+ 1` was not necessary?

Comment (muxanick) [inline javalib/src/main/scala/java/util/regex/Matcher.scala]: group(0) is source string, so actual groups start from 1 to N

Comment (muxanick) [inline javalib/src/main/scala/java/util/regex/Matcher.scala]: I'm not sure if there is one more case when re2 specifies open range as a group, like (2, inf) means the rest of the source string starts from 2th position.
There is also (inf, 5) means take(5). I have seen that kind of cases as well. But I just fixed deterministic here, I agree to fix more correctly.

**Comment by muxanick** [comment]
I don't think Travis issue is related to my changes, all green except last

**Comment by densh** [comment]
Does the suggested change address #852 ?

**Comment by muxanick** [comment]
@densh yes

**Comment by densh** [comment]
Please add a test case to make sure that #852 works now.

**Comment by muxanick** [comment]
CI errors are very strange, scripted/system-exit test failed which surely doesn't relate to this PR

**Comment by densh** [comment]
This PR needs some love from scalafmt.

**Comment by densh** [comment]
LGTM!

File: unit-tests/src/test/scala/java/util/regex/MatcherSuite.scala [hunk 2]
```diff
@@ -266,6 +264,16 @@ object MatcherSuite extends tests.Suite {
     assert(m1.lookingAt())
   }
 
+  test("issue #852, StringIndexOutOfBoundsException") {
+    val JsonNumberRegex =
+      """(-)?((?:[1-9][0-9]*|0))(?:\.([0-9]+))?(?:[eE]([-+]?[0-9]+))?""".r
+    val JsonNumberRegex(negative, intStr, decStr, expStr) = "0.000000"
+    assert(negative.isEmpty)
```
Comment (xuwei-k) [inline unit-tests/src/test/scala/java/util/regex/MatcherSuite.scala]: https://github.com/scala-native/scala-native/issues/886

**Pull request merged**